run_loop for mask rcnn first commit

MehdiShz opened the pull request:
It will alternate between 1 epoch of training and one evaluation

MehdiShz committed:
run_loop for mask rcnn first commit:
It will alternate between 1 epoch of training and one evaluation

MehdiShz committed:
run_loop for mask rcnn first commit:
It will alternate between 1 epoch of training and one evaluation

k-w-w commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,181 @@
+
+
+def main(_):
+  assert FLAGS.train_dir, '`train_dir` is missing.'
```
You can set the flag default to `None` and use `mark_flag_as_required to mark required flags. For example:
```
flags.DEFINE_string('train_dir', None, '...')
flags.mark_flag_as_required('train_dir')
```

MehdiShz replied:
why? does it work as the "assert FLAGS. ..."

k-w-w commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,181 @@
+    return dataset_util.make_initializable_iterator(
+        dataset_builder.build(config)).get_next()
+
+  train_input_dict_fn = functools.partial(get_next, train_input_config)
```
A comment here about train_input_dict_fn and eval_input_dict would be helpful. (In particular - explaining the use of `func.partial(get_next,...)`)?

MehdiShz replied:
changed to
train_input_dict_fn = get_next(train_input_config)
as the dataset_builder doesn't need another argument

k-w-w commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,181 @@
+  env = json.loads(os.environ.get('TF_CONFIG', '{}'))
+  cluster_data = env.get('cluster', None)
+  cluster = tf.train.ClusterSpec(cluster_data) if cluster_data else None
+  task_data = env.get('task', None) or {'type': 'master', 'index': 0}
```
Might be cleaner to do `task_data =  env.get('task', {'type': 'master', 'index': 0})`

k-w-w commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,181 @@
+  task_data = env.get('task', None) or {'type': 'master', 'index': 0}
+  task_info = type('TaskSpec', (object,), task_data)
+
+  # Parameters for a single worker.
```
Should we use the flag values here?

MehdiShz replied:
FLAG values will be passed to trainer.train later

k-w-w commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,181 @@
+def main(_):
+  assert FLAGS.train_dir, '`train_dir` is missing.'
+  assert FLAGS.pipeline_config_path, '`pipeline_config_path` is missing'
+  assert FLAGS.checkpoint_dir, '`checkpoint_dir` is missing.'
```
This flag has been commented out. Might be better to remove completely

MehdiShz committed:
loop added to alternate between train and eval

MehdiShz committed:
loop added to alternate between train and eval

MehdiShz committed:
omit assert for checkpoint_dir as it is the same as the training_dir

MehdiShz committed:
setting checkpoint_dir=train_dir as it is the same as the training_dir

MehdiShz committed:
num_steps flag TypeError solved

MehdiShz committed:
num_steps flag TypeError solved

MehdiShz committed:
num_steps flag TypeError solved

MehdiShz committed:
num_steps flag TypeError solved

MehdiShz committed:
num_steps flag TypeError solved

MehdiShz committed:
num_steps flag TypeError solved

MehdiShz committed:
debugging eval

MehdiShz committed:
debugging

MehdiShz committed:
debugging

MehdiShz committed:
revert

MehdiShz committed:
##Replace method "iteritems()" with "items()"
From Python 2 to 3 it seems the method "iteritems()" that is called on
Dicts has been removed. Calls have been replaced with "items()"to
restore functionality. Originally discovered this was the issue from
viewing this [here](tensorflow/models#1587)

MehdiShz committed:
##Replace method "iteritems()" with "items()"

From Python 2 to 3 it seems the method "iteritems()" that is called on
Dicts has been removed. Calls have been replaced with "items()"to
restore functionality. Originally discovered this was the issue from
viewing this [here](tensorflow/models#1587)

MehdiShz committed:
Create mask_rcnn_resnet50_atrous_coco.config

MehdiShz committed:
set reuse=tf.REUSE

MehdiShz committed:
Merge remote-tracking branch 'origin/mehdi-patch-1' into mehdi-patch-1

MehdiShz committed:
revert

MehdiShz committed:
new graph building

MehdiShz committed:
sdfsd

MehdiShz committed:
stopping_criteria_met function added for having the Option to run until
the box/mask average precision reaches the given number

MehdiShz committed:
logging added for stopping criteria

nealwu commented on object_detection/Mask_RCNN/configs/mask_rcnn_resnet50_atrous_coco.config:
```diff
@@ -0,0 +1,169 @@
+# Mask R-CNN with Resnet-50 (v1), Atrous version
```
This file is just from https://github.com/tensorflow/models/blob/master/research/object_detection/samples/configs/mask_rcnn_resnet50_atrous_coco.config right?

MehdiShz replied:
yes

nealwu commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,244 @@
+
+import tensorflow as tf
+
+EPOCHS_BETWEEN_EVALS = 1
```
We should go ahead and make this a flag as well.

nealwu commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,244 @@
+    raise Exception('eval_metrics dict does not contain the mAP field') from err
+
+  return (mask_min_ap == -1 or mask_mAP_reached > mask_min_ap) & \
+         (box_min_ap == -1 or box_mAP_reached > box_min_ap)
```
nit: the `&` should be `and` instead.

More importantly, doesn't this return true when both parameters are left unset?

nealwu commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,244 @@
+    # TODO(mehdi): make it run indef
+    total_num_epochs = 200000
+    train_config.num_steps = EPOCHS_BETWEEN_EVALS
+    total_training_cycle = total_num_epochs//train_config.num_steps
```
nit: add spaces around `//`

nealwu commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,244 @@
+# check point dir is the same as the train dir
+# flags.DEFINE_string('checkpoint_dir', '',
+#                    'Directory containing checkpoints to evaluate, typically '
+#                    'set to `train_dir` used in the training job.')
```
If we're not using this for now let's just remove it.

nealwu commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,244 @@
+def main(_):
+  assert FLAGS.train_dir, '`train_dir` is missing.'
+  assert FLAGS.pipeline_config_path, '`pipeline_config_path` is missing'
+  assert FLAGS.eval_dir, '`eval_dir` is missing.'
```
@k-w-w 's [suggestion](https://github.com/MehdiShz/reference/pull/2#discussion_r197592005) to use `mark_flag_as_required` seems like a preferable pattern here. (For some reason I can no longer reply to that comment.)

nealwu commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,244 @@
+
+  flags.DEFINE_string('checkpoint_dir', FLAGS.train_dir,
+                      'directory to read checkpoints from '
+                      'which is the same as the train_dir')
```
Better not to define a flag here separately from all the other flags. If `checkpoint_dir` is almost always the same as `train_dir`, may as well just use `train_dir` instead.

nealwu commented on object_detection/Mask_RCNN/mask_rcnn_run_loop.py:
```diff
@@ -0,0 +1,244 @@
+    ps_tasks = len(cluster_data['ps'])
+
+  if worker_replicas > 1 and ps_tasks < 1:
+    raise ValueError('At least 1 ps task is needed for distributed training.')
```
Let's spell out 'ps' here as 'parameter server'

nealwu reviewed (commented):
Looks reasonable so far. Let us know when you have a better idea of the training progress / metrics.

MehdiShz committed:
stopping condition debug

MehdiShz committed:
checkpoint flag omitted
added epochs_between_evals option
change name: ps_tasks -> parameter_server_tasks
